WalletWasabi/Blockchain/Keys/KeyManager.cs

```diff
@@ -548,169 +467,161 @@ public ExtKey GetMasterExtKey(string password)
 		}
 	}
 
-	/// <summary>
-	/// Make sure there's always clean keys generated and indexed.
-	/// Call SetMinGapLimit() to set how many keys should be asserted.
-	/// </summary>
-	public IEnumerable<HdPubKey> AssertCleanKeysIndexedAndPersist(bool? isInternal = null)
+	private void DeriveTaprootExtPubKey(ExtKey extKey)
 	{
-		var newKeys = AssertCleanKeysIndexed(isInternal);
-		if (newKeys.Any())
+		if (TaprootExtPubKey is null)
 		{
-			ToFile();
+			TaprootAccountKeyPath = GetAccountKeyPath(GetNetwork(), ScriptPubKeyType.TaprootBIP86);
+			TaprootExtPubKey = extKey.Derive(TaprootAccountKeyPath).Neuter();
 		}
-
-		return newKeys;
 	}
 
-	public IEnumerable<HdPubKey> AssertCleanKeysIndexed(bool? isInternal = null)
+	public void SetKeyState(KeyState newKeyState, HdPubKey hdPubKey)
 	{
-		var newKeys = new List<HdPubKey>();
+		if (hdPubKey.KeyState == newKeyState)
+		{
+			return;
+		}
 
-		if (isInternal.HasValue)
+		hdPubKey.SetKeyState(newKeyState);
+		if (newKeyState is KeyState.Locked or KeyState.Used)
 		{
-			while (CountConsecutiveUnusedKeys(isInternal.Value, ignoreTail: false) < MinGapLimit)
+			var keySource = GetHdPubKeyGenerator(hdPubKey.IsInternal, hdPubKey.FullKeyPath.GetScriptTypeFromKeyPath());
+
+			// This can happen after downgrading to pre-taproot wasabi version the switching back to a supporting
+			// version so taproot keys are detected. However, the user has not login yet so taprootextpubkey is
+			// not derived yet (because pre-taproot wasabi do not serialize fields that it doesn't know)
+			if (keySource is { })
 			{
-				newKeys.Add(GenerateNewKey(SmartLabel.Empty, KeyState.Clean, isInternal.Value));
+				var view = HdPubKeyCache.GetView(keySource.KeyPath);
+				HdPubKeyCache.AddRangeKeys(keySource.AssertCleanKeysIndexed(view).Select(CreateHdPubKey));
 			}
 		}
-		else
+	}
+
+	private HdPubKeyGenerator? GetHdPubKeyGenerator(bool isInternal, ScriptPubKeyType scriptPubKeyType) =>
```

Comment:
The return value should not be Nullable.

Reply:
Yes. I was tempted to throw an exception but then, given this is called many times during transaction processing, I changed it to `null` (i hate it) otherwise it could hit the loading time even more. Anyway, I didn't measure it and i assume it can have a negative impact but I really don't know.